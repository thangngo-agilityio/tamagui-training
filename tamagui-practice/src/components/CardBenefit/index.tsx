import { Box, Flex, Heading, Text } from '@chakra-ui/react';
import { memo, ReactElement } from 'react';
import isEqual from 'react-fast-compare';

type TCardBenefit = {
  icon?: ReactElement;
  title?: string;
  text?: string;
};

const CardBenefit = ({ icon, title, text }: TCardBenefit) => (
  <Flex alignItems="center" flexDirection="row">
    <Box>{icon}</Box>
    <Flex maxW="206px" ml="24px" flexDirection="column">
      <Heading size="xl" mb="2px">
        {title}
      </Heading>
      <Text variant="senary" size="textLg">
        {text}
      </Text>
    </Flex>
  </Flex>
);

export default memo(CardBenefit, isEqual);
