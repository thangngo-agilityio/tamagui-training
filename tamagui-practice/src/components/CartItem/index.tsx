import Image from 'next/image';
import { Box, Flex, Heading, Text, VStack } from '@chakra-ui/react';
import { useCallback } from 'react';

// Utils
import { formatAmountNumber } from '@/utils';

// Icons
import { DeleteIcon } from '@/icons';

type TCartItem = {
  id: string;
  title: string;
  image: string[];
  description: string;
  price: number;
  quantity: number;
  onRemoveItem?: (id: string) => void;
};

const CartItem = ({
  id,
  title,
  image,
  description,
  price,
  quantity,
  onRemoveItem,
}: TCartItem) => {
  const handleRemoveItem = useCallback(() => {
    onRemoveItem && onRemoveItem(id);
  }, [id, onRemoveItem]);
  return (
    <Flex flexDir={{ base: 'column', lg: 'row' }} gap="30px">
      <Flex flex={1}>
        <Image
          width={347}
          height={290}
          src={image[0]}
          alt={title}
          style={{
            objectFit: 'cover',
            width: '100%',
            height: '100%',
            borderRadius: '8px',
          }}
        />
      </Flex>
      <Flex flex={4} flexDir="column">
        <Heading
          size={{ base: 'size4xl', lg: 'size6xl' }}
          variant="quinary"
          mb="5px"
        >
          {title}
        </Heading>
        <Text
          size={{ base: 'textMd', lg: 'text2Xl' }}
          variant="senary"
          mb="30px"
        >
          {description}
        </Text>

        <VStack mb="30px" alignItems="flex-start">
          <Heading variant="productTitle" size="size2xl">
            Size:
          </Heading>
          <Flex flexDir="row" gap="5px">
            <Text variant="senary" size="text2Xl">
              Length-34cm,
            </Text>
            <Text variant="tertiary" size="text2Xl">
              Width-56cm
            </Text>
          </Flex>
        </VStack>

        <Box w="50%" borderTopWidth="1px" borderColor="border.300" mb="10px">
          <Heading size="size9xl" variant="quinary" py="10px">
            N{formatAmountNumber(price?.toString())}
          </Heading>
        </Box>

        <Flex gap="10px">
          <Heading variant="productTitle" size="size2xl">
            Quantity:
          </Heading>
          <Text variant="tertiary" size="text2Xl">
            {quantity}
          </Text>
        </Flex>
      </Flex>

      <Flex justifyContent="center" alignItems="center">
        <Box height="fit-content" onClick={handleRemoveItem} cursor="pointer">
          <DeleteIcon />
        </Box>
      </Flex>
    </Flex>
  );
};

export default CartItem;
