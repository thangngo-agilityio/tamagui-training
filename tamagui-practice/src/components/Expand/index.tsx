import {
  Box,
  Drawer,
  DrawerBody,
  DrawerCloseButton,
  DrawerContent,
  DrawerHeader,
  DrawerOverlay,
  Flex,
  Heading,
  HStack,
  Link,
  Stack,
  Text,
  useDisclosure,
  VStack,
} from '@chakra-ui/react';

// Component
import {
  CartIcon,
  HamburgerIcon,
  HeartIcon,
  LineIcon,
  LogoIcon,
  Logout,
} from '@/icons';

// Constants
import { ROUTER } from '@/constants';

type TSidebarProps = {
  totalQuantity: number;
  onClick?: () => void;
};

const ExpandSidebar = ({ totalQuantity, onClick }: TSidebarProps) => {
  const { isOpen, onOpen, onClose } = useDisclosure();
  return (
    <Stack>
      <Box
        onClick={onOpen}
        top="35px"
        left="10px"
        data-testid="button-hamburger"
      >
        <HamburgerIcon />
      </Box>
      <Drawer
        isOpen={isOpen}
        placement="left"
        onClose={onClose}
        data-testid="button-close"
      >
        <DrawerOverlay />
        <DrawerContent alignItems="center" justifyContent="center">
          <DrawerCloseButton />
          <DrawerHeader>
            <Link href="/" as="h1" mb="27.5px">
              <LogoIcon />
            </Link>
            <LineIcon />
          </DrawerHeader>

          <DrawerBody w="100%" px="55px">
            <VStack gap="15px" alignItems="flex-start">
              <HStack
                w="100%"
                borderBottom="1px solid"
                borderColor="border.500"
                pb="10px"
              >
                <HeartIcon />
                <Heading>Favorites</Heading>
              </HStack>

              <HStack
                as={Link}
                href={ROUTER.CART}
                w="100%"
                borderBottom="1px solid"
                borderColor="border.500"
                pb="10px"
              >
                <Box
                  position="relative"
                  opacity={1}
                  transition=".2s ease-in"
                  _hover={{ opacity: '.8' }}
                >
                  <CartIcon />
                  <Flex
                    w="20px"
                    h="20px"
                    bgColor="background.1800"
                    borderRadius="100%"
                    alignItems="center"
                    justifyContent="center"
                    position="absolute"
                    top={-3}
                    right={-3}
                  >
                    <Text size="textMd" variant="secondary">
                      {totalQuantity}
                    </Text>
                  </Flex>
                </Box>
                <Heading>Cart</Heading>
              </HStack>

              <HStack
                w="100%"
                borderBottom="1px solid"
                borderColor="border.500"
                pb="10px"
                onClick={onClick}
                cursor="pointer"
                data-testid="button-logout"
              >
                <Logout />
                <Heading>Logout</Heading>
              </HStack>
            </VStack>
          </DrawerBody>
        </DrawerContent>
      </Drawer>
    </Stack>
  );
};

export default ExpandSidebar;
