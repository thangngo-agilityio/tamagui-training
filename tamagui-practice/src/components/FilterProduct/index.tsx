'use client';

import { Fragment, useCallback } from 'react';
import { Flex, RadioGroup } from '@chakra-ui/react';

// Components
import ItemCategory from '../ItemCategory';

// Constants
import { MENU_ITEM_FILTER, SEARCH_QUERIES } from '@/constants';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import { getSearchParams, updateSearchParams } from '@/utils';

const FilterProduct = () => {
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const { replace } = useRouter();
  const { category = '' } = getSearchParams(searchParams);

  const handleFilterProduct = useCallback(
    (category: string) => {
      const value = category;

      const updatedParams = updateSearchParams(
        searchParams,
        SEARCH_QUERIES.CATEGORIES,
        value,
      );

      replace(`${pathname}?${updatedParams.toString()}`, { scroll: false });
    },
    [pathname, replace, searchParams],
  );
  return (
    <Flex
      width={{ base: '100%', lg: '73%' }}
      px={{ base: '28px', lg: '145px' }}
      pt="20px"
      pb="26px"
      bgColor={{ base: 'transparent', lg: 'background.800' }}
      boxShadow={{ base: 'unset', lg: '0 11px 30px 4px rgba(0, 0, 0, 7%)' }}
      gap={{ base: '25px', lg: '75px' }}
      justifyContent="center"
      alignItems="center"
      mt={{ base: 'unset', lg: '-20px' }}
      mb="146px"
    >
      {MENU_ITEM_FILTER.map((item) => {
        const IconComponent = item.icon || Fragment;
        const IconActiveComponent = item.iconActive || Fragment;
        return (
          <RadioGroup
            flexDirection="column"
            alignItems="center"
            cursor="pointer"
            onChange={handleFilterProduct}
            value={category}
            key={item.id}
          >
            <ItemCategory
              value={item.value}
              title={item.itemContent}
              filter={category}
              icon={<IconComponent />}
              iconActive={<IconActiveComponent />}
            />
          </RadioGroup>
        );
      })}
    </Flex>
  );
};

export default FilterProduct;
