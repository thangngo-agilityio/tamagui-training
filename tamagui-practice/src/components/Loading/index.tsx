// Libs
import { Spinner as SpinnerChakra, Center } from '@chakra-ui/react';

const LoadingIndicator = () => (
  <Center minHeight="100vh">
    <SpinnerChakra />
  </Center>
);

export default LoadingIndicator;
