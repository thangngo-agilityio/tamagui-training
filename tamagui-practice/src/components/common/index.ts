export { default as InputField } from './InputFiled';
export { default as Avatar } from './Avatar';
