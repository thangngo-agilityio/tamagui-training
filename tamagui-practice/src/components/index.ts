export * from './Skeleton';
export { default as LoginForm } from './Form/Login';
export { default as SignUpForm } from './Form/SignUp';
export { default as ItemCategory } from './ItemCategory';
export { default as CardBenefit } from './CardBenefit';
export { default as ProductCard } from './ProductCard';
export { default as NavFooter } from './NavFooter';
export { default as UserDropdown } from './UserDropdown';
export { default as Loading } from './Loading';
export { default as LoadingIndicator } from './LoadingIndicator';
export { default as Pagination } from './Pagination';
export { default as SkeletonProductList } from './Skeleton/ProductList';
export { default as SkeletonProductDetail } from './Skeleton/ProductDetail';
export { default as SkeletonSection } from './Skeleton/Section';
export { default as SkeletonCartList } from './Skeleton/CartList';
export { default as ImageProduct } from './ImageProduct';
export { default as ProductInfo } from './ProductInfo';
export { default as CartItem } from './CartItem';
export { default as TotalCart } from './TotalCart';
export { default as OrderSuccess } from './OrderSuccess';
export { default as Expand } from './Expand';
export { default as InputSearch } from './InputSearch';
export { default as Navigation } from './Navigation';
export { default as FilterProduct } from './FilterProduct';
export { default as ProductList } from './ProductList';
export { default as CartList } from './CartList';

// Common
export * from './common';
export { default as HeadingSection } from './common/HeadingSection';
