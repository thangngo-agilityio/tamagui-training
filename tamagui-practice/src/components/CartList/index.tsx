'use client';

import { Box, Flex } from '@chakra-ui/react';
import { useCallback } from 'react';

// Actions
import { updateMyCart } from '@/actions';

// Components
import { CartItem } from '@/components';

// Hooks
import { useCustomToast } from '@/hooks';

// Constants
import { SUCCESS_MESSAGES } from '@/constants';

// Types
import { ICartItem } from '@/types';

type TCartList = {
  cartItem: ICartItem[];
  cartId: string;
};

const CartList = ({ cartItem, cartId }: TCartList) => {
  const { showToast } = useCustomToast();

  const handleRemoveItem = useCallback(
    async (productId: string) => {
      const newCartItems = cartItem.filter((Item) => {
        const { product } = Item || {};
        const { id } = product || {};

        return id !== productId;
      });
      const res = await updateMyCart(cartId, newCartItems);

      const { error } = res || {};

      if (error) {
        showToast(error);
      } else {
        showToast(SUCCESS_MESSAGES.DELETE_CART_ITEM, 'success');
      }
    },
    [cartId, cartItem, showToast],
  );
  return (
    <Flex flexDir="column" mb="90px">
      {cartItem.map((cart) => {
        const { product, quantity } = cart || {};
        const {
          id = '',
          name = '',
          image = [],
          description = '',
          price = 0,
        } = product || {};

        const total = price * quantity;
        return (
          <Box
            mb="36px"
            pb="54px"
            borderBottom="1px solid"
            borderColor="border.500"
            _last={{
              mb: 'unset',
              pb: 'unset',
              borderBottom: 'unset',
              borderColor: 'unset',
            }}
            key={id}
          >
            <CartItem
              id={id}
              title={name}
              image={image}
              description={description}
              price={total}
              quantity={quantity}
              onRemoveItem={handleRemoveItem}
            />
          </Box>
        );
      })}
    </Flex>
  );
};

export default CartList;
