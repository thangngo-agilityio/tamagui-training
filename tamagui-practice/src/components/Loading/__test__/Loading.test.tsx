import { Loading } from '@/components';
import { render } from '@testing-library/react';

describe('Loading component', () => {
  it('Should match snapshot for Loading', () => {
    const { container } = render(<Loading />);

    expect(container).toMatchSnapshot();
  });
});
