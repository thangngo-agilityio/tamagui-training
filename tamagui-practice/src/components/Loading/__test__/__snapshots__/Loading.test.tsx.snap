// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading component Should match snapshot for Loading 1`] = `
<div>
  <div
    class="css-1o0ed15"
  >
    <div
      class="chakra-spinner css-1y7joxr"
    >
      <span
        class="css-8b45rq"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
